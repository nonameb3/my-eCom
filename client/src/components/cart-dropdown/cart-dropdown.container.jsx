import React from "react";
import { useQuery, useMutation } from "react-apollo";
import { gql } from "apollo-boost";

import CartDropdownCompoent from "./cart-dropdown";

const TOGGLE_CART_HIDDENT = gql`
  mutation TogggleCartHidden {
    toggleCartHidden @client
  }
`;

const GET_CART_ITEMS = gql`
  {
    cartItems @client
  }
`;

export default function() {
  const {data} = useQuery(GET_CART_ITEMS);
  const mutation = useMutation(TOGGLE_CART_HIDDENT);

  return (
    <CartDropdownCompoent
      cartItems={data.cartItems}
      toggleCartHidden={mutation[0]}
    />
  );
}
