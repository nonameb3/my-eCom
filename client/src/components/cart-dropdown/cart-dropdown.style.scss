.cart-dropdown {
  position: absolute;
  width: 240px;
  height: 340px;
  display: flex;
  flex-direction: column;
  padding: 20px;
  border: 1px solid black;
  background-color: white;
  top: 90px;
  right: 40px;
  z-index: 5;

  .cart-items {
    height: 240px;
    display: flex;
    flex-direction: column;
    overflow: hidden;
    background: transparent;

    .empty-text {
      font-size: 18px;
      font-weight: bold;
      margin: 50px auto;
    }
    
    // custom scrollbar
    &:hover {
      overflow: overlay;
    }
  
    &::-webkit-scrollbar {
      width: 10px;
      height: 0px;
      background: transparent;
    }
  
    &::-webkit-scrollbar-track {
      border-radius: 10px;
      background-color: #f5f5f5;
    }
  
    &::-webkit-scrollbar-thumb {
      border-radius: 10px;
      background-color: #555;
    }
  }

  button {
    margin-top: auto;
  }
}
