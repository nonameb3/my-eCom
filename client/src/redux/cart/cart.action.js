import * as TYPE from "./cart.type";

export const toggleCart = () => {
  return {
    type: TYPE.TOGGLE_CART_DROPDOWN
  };
};

export const addCartItem = item => {
  return {
    type: TYPE.ADD_CART_ITEMS,
    payload: item
  };
};

export const deleteCartItem = item => {
  return {
    type: TYPE.DELETE_CART_ITEM,
    payload: item
  };
};

export const removeCartItem = item => {
  return {
    type: TYPE.REMOVE_CART_ITEM,
    payload: item
  };
};

export const clearAllCartItems = () => {
  return { 
    type: TYPE.CLEAR_CART
  }
};
