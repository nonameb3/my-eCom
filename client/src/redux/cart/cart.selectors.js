import { createSelector } from "reselect";

// reselect will select only value change
const selectCart = state => state.cart;

export const selecterCartItems = createSelector(
  [selectCart],
  cart => cart.cartItems
);

export const selectCartHidden = createSelector(
  [selectCart],
  cart => cart.hidden
);

export const selecterCartItemsCount = createSelector(
  [selecterCartItems],
  cartItems => cartItems.reduce((preItem, item) => preItem + item.quantity, 0)
);

export const selecterCartItemsTotalCount = createSelector(
  [selecterCartItems],
  cartItems =>
    cartItems.reduce((preItem, item) => preItem + item.quantity * item.price, 0)
);
