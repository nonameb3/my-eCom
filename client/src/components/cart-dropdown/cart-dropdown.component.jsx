import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { compose } from "redux";
import { withRouter } from "react-router-dom";

import CartItem from "../cart-item/cart-item.component";
import CustomButton from "../custom-button/custom-button.component";
import { toggleCart } from "../../redux/cart/cart.action";
import { selecterCartItems } from "../../redux/cart/cart.selectors";
import "./cart-dropdown.style.scss";

export const CartDropdown = ({ history }) => {
  const cartItems = useSelector(state => selecterCartItems(state));
  const dispatch = useDispatch();

  const onClick = () => {
    history.push("/checkout");
    dispatch(toggleCart());
  };

  return (
    <div className="cart-dropdown">
      <div className="cart-items">
        {cartItems.length ? (
          cartItems.map(cartItem => (
            <CartItem key={cartItem.id} item={cartItem} />
          ))
        ) : (
          <span className="empty-text">Cart is Empty</span>
        )}
      </div>
      <CustomButton onClick={onClick}>GO TO CHECKOUT</CustomButton>
    </div>
  );
};

export default compose(withRouter)(CartDropdown);
