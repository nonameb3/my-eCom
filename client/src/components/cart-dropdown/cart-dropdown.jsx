import React from "react";
import { connect } from "react-redux";
import { compose } from "redux";
import { withRouter } from "react-router-dom";

import CartItem from "../cart-item/cart-item.component";
import CustomButton from "../custom-button/custom-button.component";
import { selecterCartItems } from "../../redux/cart/cart.selectors";
import "./cart-dropdown.style.scss";

function cartDropdown({ cartItems, history }) {
  return (
    <div className="cart-dropdown">
      <div className="cart-items">
        {cartItems.length ? (
          cartItems.map(cartItem => (
            <CartItem key={cartItem.id} item={cartItem} />
          ))
        ) : (
          <span className="empty-text">Cart is Empty</span>
        )}
      </div>
      <CustomButton onClick={() => history.push("/checkout")}>
        GO TO CHECKOUT
      </CustomButton>
    </div>
  );
}

const mapStateToProps = state => {
  return { cartItems: selecterCartItems(state) };
};

export default compose(
  withRouter,
  connect(mapStateToProps)
)(cartDropdown);
