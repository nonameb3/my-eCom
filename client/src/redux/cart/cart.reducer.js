import * as TYPE from "./cart.type";
import { addNewCartItems, removeCartItem } from './cart.utils'

const INITIAL_STATE = {
  hidden: true,
  cartItems: []
};

export default function(state = INITIAL_STATE, action) {
  switch (action.type) {
    case TYPE.TOGGLE_CART_DROPDOWN:
      return {
        ...state,
        hidden: !state.hidden
      };
    case TYPE.ADD_CART_ITEMS:
      return {
        ...state,
        cartItems: addNewCartItems(state.cartItems, action.payload)
      }
    case TYPE.REMOVE_CART_ITEM:
      return {
        ...state,
        cartItems: removeCartItem(state.cartItems, action.payload)
      }
    case TYPE.DELETE_CART_ITEM:
      return {
        ...state,
        cartItems: state.cartItems.filter(item => item.id !== action.payload.id)
      }
    case TYPE.CLEAR_CART:
      return {
        ...state,
        cartItems: []
      }
    default:
      return state;
  }
}
