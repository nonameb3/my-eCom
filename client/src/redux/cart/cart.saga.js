import { all ,put ,call ,takeLatest } from 'redux-saga/effects';

import { clearAllCartItems } from './cart.action';
import * as USERTYPE from '../user/user.type';

export function* clearCartStore() {
  yield put(clearAllCartItems());
}

// handle function
export function* onSingOutSuccess() {
  yield takeLatest(USERTYPE.SIGN_OUT_SUCCESS, clearCartStore)
}

export function* cartSaga() {
  yield all([call(onSingOutSuccess)])
}
