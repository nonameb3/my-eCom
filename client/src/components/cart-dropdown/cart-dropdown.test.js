import React from "react";
import { shallow } from "enzyme";

import { CartDropdown } from "./cart-dropdown.component";
import CustomButton from "../custom-button/custom-button.component";
import CartItem from "../cart-item/cart-item.component";
import { toggleCart } from "../../redux/cart/cart.action";

let wrapper;
const mockDispatch = jest.fn();
let mockHistory = { push: jest.fn() };
let mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

jest.mock("react-redux", () => ({
  useSelector: jest.fn(() => mockCartItems),
  useDispatch: () => mockDispatch
}));

const mockProps = {
  history: mockHistory
};

describe("CartDropdown", () => {
  beforeEach(() => {
    wrapper = shallow(<CartDropdown {...mockProps} />);
  });

  it("should rander CartDropdown", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("shouild rander cartItem same as array", () => {
    expect(wrapper.find(CartItem).length).toBe(mockCartItems.length);
  });

  it('should call history.push when button is clicked', () => {
    wrapper.find(CustomButton).simulate('click');
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCart());
  });
});

it("should render empty-message if cartItems is empty", () => {
  mockCartItems = [];
  wrapper = shallow(<CartDropdown {...mockProps} />);
  expect(wrapper.find(".empty-text")).toHaveLength(1);
});
