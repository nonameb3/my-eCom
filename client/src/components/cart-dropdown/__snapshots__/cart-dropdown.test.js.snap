// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown should rander CartDropdown 1`] = `ShallowWrapper {}`;
